embassy-stm32f4/src/rtc.rs

```diff
@@ -0,0 +1,290 @@
+use core::cell::Cell;
+use core::mem;
+use core::ops::Deref;
+use core::sync::atomic::Ordering;
+use embassy::time::Clock;
+
+use crate::hal::bb;
+use crate::hal::rcc::Clocks;
+use crate::interrupt;
+use crate::interrupt::{CriticalSection, Mutex, OwnedInterrupt};
+use crate::pac::{tim6, RCC};
+
+struct AlarmState {
+    timestamp: Cell<u64>,
+    callback: Cell<Option<(fn(*mut ()), *mut ())>>,
+}
+
+impl AlarmState {
+    fn new() -> Self {
+        Self {
+            timestamp: Cell::new(u64::MAX),
+            callback: Cell::new(None),
+        }
+    }
+}
+
+const ALARM_COUNT: usize = 3;
+
+pub struct RTC<T: Instance> {
+    rtc: T,
+    irq: T::Interrupt,
+    // Timestamp marks the elapsed time, excluding the current timer.
+    timestamp: u64,
+    clocks: Clocks,
+
+    /// Timestamp at which to fire alarm. u64::MAX if no alarm is scheduled.
+    alarms: Mutex<[AlarmState; ALARM_COUNT]>,
+}
+
+unsafe impl<T: Instance> Send for RTC<T> {}
+unsafe impl<T: Instance> Sync for RTC<T> {}
+
+impl<T: Instance> RTC<T> {
+    pub fn new(rtc: T, irq: T::Interrupt, clocks: Clocks) -> Self {
+        T::enable_clock();
+
+        Self {
+            rtc,
+            irq,
+            timestamp: 0,
+            clocks: clocks,
+            alarms: Mutex::new([AlarmState::new(), AlarmState::new(), AlarmState::new()]),
+        }
+    }
+
+    #[inline(always)]
+    fn reset_timestamp(&mut self) {
+        self.timestamp += self.rtc.deref().arr.read().bits() as u64;
+    }
+
+    #[inline(always)]
+    fn set_arr(&self, arr: u32) {
+        if arr != self.rtc.deref().arr.read().bits() {
+            self.rtc.deref().arr.write(|w| unsafe { w.bits(arr) });
+            self.update();
+        }
+    }
+
+    #[inline(always)]
+    fn update(&self) {
+        self.rtc.deref().cr1.modify(|_, w| w.urs().set_bit());
+        self.rtc.deref().egr.write(|w| w.ug().set_bit());
+        self.rtc.deref().cr1.modify(|_, w| w.urs().clear_bit());
+    }
+
+    pub fn start(&mut self) {
+        // stop counter
+        self.rtc.deref().cr1.modify(|_, w| w.cen().clear_bit());
+        // reset counter
+        self.rtc.deref().cnt.reset();
+
+        let frequency = 1; // timeout.into().0;
+        let pclk_mul = if T::ppre(self.clocks) == 1 { 1 } else { 2 };
+        let ticks = T::pclk(self.clocks) * pclk_mul / frequency;
+
+        let psc = ((ticks - 1) / (1 << 16)) as u16;
+        self.rtc.deref().psc.write(|w| w.psc().bits(psc));
+
+        self.set_arr(u16::MAX as u32);
```

Comment:
Could you explain how this achieve 32768 ticks per second ?

Reply:
This is copied from stm32f4xx-hal, and it seems to work after extensive testing. 

https://github.com/stm32-rs/stm32f4xx-hal/blob/master/src/timer.rs#L307

Reply:
But there these calculations are used (together with the ARR) to provide one interrupt at a certain frequency, not sure how that translates to here, since you aren't even using the ARR and a frequency of 1.